lib/full-site-editing/block-templates.php

```diff
@@ -359,7 +359,28 @@ function gutenberg_get_block_template( $id, $template_type = 'wp_template' ) {
 	$template_query       = new WP_Query( $wp_query_args );
 	$posts                = $template_query->get_posts();
 
+	/**
+	 * Check that we have retrieved the correct post.
+	 * When a post doesn't exist that matches the above tax_query, the WP_Query can
+	 * return a post matching the other above parameters (slug, post_type, etc.).
+	 * This means if we have a template saved for "Theme-A", if we switch to "Theme-B"
```

Comment:
I'm thinking that providing the slug and post type above is making the query too specific, so it doesn't make sense for the query to filter further by taxonomy and that part is just dropped (as pointed out in @david-szabo97's SQL query log).

Could we just remove the `tax_query` from the args, and perform the check on the the returned post afterwards. Instead of the code below we could `has_term` (e.g. `has_term( $theme, 'wp_theme', $posts[0] )`).

Reply:
Wouldn't this omit posts if we have more than one result (multiple themes with the same template)

Reply:
It would, by my understanding was that the above tax query was trying to achieve that anyway? (filter out posts associated with other themes)

Reply:
> I'm thinking that providing the slug and post type above is making the query too specific,

Correct, thats what I was trying to say above with all the code links.  If `name` is provided, the tax related code in the query class doesn't run.  

> Could we just remove the tax_query from the args, and perform the check on the the returned post afterwards.

With the name field marking the query as singular, I would expect that to only retrieve 1 post?  The problem is we will have X amount of posts with the same name (like 'header' or 'index'), so just checking the first that is returned wont give us an accurate answer.  If it DOES return multiple, we can loop through all results... but this also doesn't seem friendly in terms of scalability?

Similarly, we could drop the name and query everything based on post type and tax term.  But then we would need to loop all results to see if the specific 'name' is in there.  And that would have the same concerns with performance and scalability?

Reply:
I updated the comment in the code here as well as the PR description to better reflect the more recent findings.  But commenting my main questions on how we move forward:

> Do we find a safe way to update the query class in core? Do we adapt our queries here to be less specific and then loop through all results to see if the post in question exists? Do we find a new way to save the theme value for templates/parts that is not taxonomy?

Reply:
Heh, it looks like we can get around all of this by using `post_name__in` instead of `name`.  There were a handful of other issues as well regarding the unique_post_slug function, but I think we have things working here. (more in my comments below)

Reply:
> The problem is we will have X amount of posts with the same name (like 'header' or 'index'), so just checking the first that is returned wont give us an accurate answer. If it DOES return multiple, we can loop through all results... but this also doesn't seem friendly in terms of scalability?

I wrongly assumed that we'll have unique slug values for some reason, but after testing this out a bit I realized that that's not the case. As for looping through all results and scalability - it shouldn't be an issue given that it would be iterating over on the order of tens of objects in memory (even if it was more, DB query is still more expensive). But that besides the point now, since the new approach looks better. :)

> Heh, it looks like we can get around all of this by using post_name__in instead of name

Nice!